### DISCUSSION TITLE
Nessus automatic scan upload with Nessus Prof. API

### DISCUSSION THREAD

**Discussion by QuaiGoner** [DefectDojo/django-DefectDojo#5195] [Q&A] [answered]
Hello. How hard would it to implement importing Nessus Scan Results with an Nessus API in real-time?

**Comment by damiencarol** [comment]
It depends on the complexity of the API, if we had a Python module ready to use the API.
Also it could help if the data of the API is close to what we have in the XML export, as we will be able to reuse some code.

**Reply by StefanFl** [reply]
We should do some structural work before, so we have a flexible mechanism to manage the configuration for accessing the Nessus API (url, credentials). This is on my tasklist for the next month.

**Reply by QuaiGoner** [reply]
I have working (partially) pipeline with nessus + DD in postman will it help to integrate those in DD?

**Reply by damiencarol** [reply]
It's basically converting your pipeline to Python code.
@StefanFl some structural work has been done around configuration. Do you think we can take another look to this question?

**Reply by StefanFl** [reply]
The configuration should be flexible enough now. If someone has access to a Nessus installation and is willing to build the API integration, they would be good to go. The integrations for SonarQube and Cobalt.io are good starting points.

**Reply by damiencarol** [reply]
@QuaiGoner are you ready to work on something like that? We can help/provide support.

**Reply by Anillaerdal** [reply]
Hi All,

Iam also looking for DefectDojo Automatic scan upload with Nessus API.
Please share me how to config Defectdojo with Nessus API ?

**Reply by blackaichi** [reply]
Hi all, 

Any update on this? If not perhaps I can help you

**Reply by Anillaerdal** [reply]
> Hi all,
> 
> Any update on this? If not perhaps I can help you

Hi Still no one reply on this, please help on this it is very help full for me.

**Reply by Stamka** [reply]
Hi, In our company we developed a python script that it parses folders in nessusPro by api than upload reports in csv to defectdojo by crontab

**Reply by QuaiGoner** [reply]
@Stamka hello! How do you get around duplicates and previously made automatic Jira issues?
Do you upload new scan results in the same engagement with close old finding option?

**Reply by Stamka** [reply]
@QuaiGoner Our engineers and devops working directly in defectodojo without jira. 

> Do you upload new scan results in the same engagement with close old finding option?

I create new engagement and do deduplication and autoclosing old findings for the whole project

**Comment by manuel-sommer** [comment, marked as answer]
Hi all,

I guess an additional API parser will not be integrated / accepted as a PR from DefectDojo side. The reason can be found here:
https://defectdojo.github.io/django-DefectDojo/contributing/how-to-write-a-parser/#api-parsers

"DefectDojo has a limited number of API parsers. While we won’t remove these connectors, adding API connectors has been problematic and thus we cannot accept new API parsers / connectors from the community at this time for supportability reasonsing. To maintain a high quality API connector, it is necessary to have a license to the tool. To get that license requires partnership with the author or vendor. We’re close to announcing a new program to help address this and bring API connectors to DefectDojo."

Thus, I would recommend to build your own automation to upload the files to DefectDojo or use the official support.

Best,

Manuel

**Comment by Noonster77** [comment]
![dojotest11 py](https://github.com/user-attachments/assets/7debfd8d-2647-4208-80e3-d5a1b91449d0)
Here is my custom dojo script. It connects to a tenable.sc instance and allows you to export queries into a .csv file, then automatically imports the findings into a defect dojo project, and engagement with the same key words as the tenable.sc query. It will reimport the findings under the same test so it keeps track of how many were opened or closed.

**Reply by stanleylin75** [reply]
Hello! @Noonster77 
would you please share the python script again, 
it's really helpful
thanks

**Reply by manuel-sommer** [reply]
Hi @stanleylin75 ,
you just have to download this python skript (sojotest11.py) and save it as .py, not as .jpeg or as an image format.

**Comment by Noonster77** [comment]
Here is my updated script. It has all the variables at the top for easy use. How its used. 1) Go into security center and goto vulnerabilty analysis, apply your filters (I used, assets, plugin type active, severity (low, moderate, high, critical) and last observed in last 7 days). From there I can save the query. I usually call it Project - RHEL8 - Vulnerabilites etc(note the script has project prefix, which means it will only pull queries starting with that project name. Make sure you set the tags in the query(use as many as you want). Then go into defect dojo and create the product with the same exact tags. --config selects which queries you want to run, --config-test allows you to run a export/import for testing purposes, and --export runs all the queries that were saved. It queries tenable.sc exports a csv report of the findings, queries defect dojo, finds the product, checks if there is an existing enagement for the tenable scan and if not creates it. I run the script once a week, and it automatically finds the latest tenable scan/engagement and reimports the data. So you can track items that were closed or new findings. It also send you a summary email of all the exports, md5 hash,  how many IP's were scanned, and any errors. I recommend one query per product. (also seperate product types from vulnerability and compliance. This will automate the weekly scan results into defect dojo for tracking. Let me know if you have any questions. 
![dojo-export.py (1).txt](https://github.com/user-attachments/files/19509300/dojo-export.py.1.txt)


[]()

**Reply by stanleylin75** [reply]
thanks, it works.

and there are 2 tags in Tenable.sc, but import to the same product. 

any idea :(

**Comment by Noonster77** [comment]
The script should handle as many tags as you need. The queries and the
products have to have the same tags to match.
Thanks!

On Mon, Mar 31, 2025 at 5:02 AM stanleylin75 ***@***.***>
wrote:

> thanks, it works.
>
> and there are 2 tags in Tenable.sc, but import to the same product.
>
> any idea :(
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_DefectDojo_django-2DDefectDojo_discussions_5195-23discussioncomment-2D12674056&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=ALyql_YvQCxwa_vRDQY02RnxMu2j2Os3Cl6QF8vOmRHS1wZMLnIkeeQfUWXjwdpB&s=Sm4O6HKai1WYytu7TUizTK_GKMmfvShYxS5khjqnJW8&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_BCGFMJB2TYBPK3LD66Y46S32XEACBAVCNFSM6AAAAABKQGMJP6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENRXGQYDKNQ&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=ALyql_YvQCxwa_vRDQY02RnxMu2j2Os3Cl6QF8vOmRHS1wZMLnIkeeQfUWXjwdpB&s=nBZF53OtwItNCLtoXKA1mOb2fMYCjlMqi0Ox_E-QRpU&e=>
> .
> You are receiving this because you were mentioned.Message ID:
> <DefectDojo/django-DefectDojo/repo-discussions/5195/comments/12674056@
> github.com>
>


-- 

Click here for "Request Forms"
<https://sites.google.com/verizon.com/publicsectorcso-scan/request-forms>


*Josh Noonan – CISSP, CEH, ECSA*

Public Sector CyberSecurity Organization

Verizon

703-728-2733

**Comment by stanleylin75** [comment]
there are 4 queries in Tenable.SC, but there is only 1 i can select when running dojo-export --config

![image](https://github.com/user-attachments/assets/6e888256-64fa-479f-a952-886fc06c4f6e)

and 2 tags in tenable.sc, and product also, but import to the 1 product 
![image](https://github.com/user-attachments/assets/cdb53b79-dd2c-4981-ba5f-a12a4eec3609)

**Comment by Noonster77** [comment]
1) Make sure in the dojo.py file the project name you have at the beginning
of your queries. For example if I put in ABC, then my queries need to start
with ABC.
2)I saw the workflow, and it exported the csv file with no tags. (Make sure
the query itself has at least 2 tags, and then the matching product in
defect dojo has at least two tags.

My setup:

(query 1) abc-rhel7-vuln
keywords abc, rhel7, vuln.

(Product 1) ABC - Rhel7 - Vulnerabilities
keywords abc, rhel7, vuln

Thanks!

Josh

On Wed, Apr 2, 2025 at 3:40 AM stanleylin75 ***@***.***>
wrote:

> there are 4 queries in Tenable.SC, but there is only 1 i can select when
> running dojo-export --config
>
> image.png (view on web)
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_user-2Dattachments_assets_6e888256-2D64fa-2D479f-2Da952-2D886fc06c4f6e&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=aNxHlfDsWx5OdvjtsyfULQ1oCWePGPJ4WOrfMYHzu2-UN2lb791PQcnDh5QeqPTr&s=IId9zN_1bc_1xQIwvSR4MHwWURhXo2L2bSmNurRUetM&e=>
>
> and 2 tags in tenable.sc, and product also, but import to the 1 product
> image.png (view on web)
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_user-2Dattachments_assets_cdb53b79-2Ddd2c-2D4981-2Dba5f-2Da12a4eec3609&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=aNxHlfDsWx5OdvjtsyfULQ1oCWePGPJ4WOrfMYHzu2-UN2lb791PQcnDh5QeqPTr&s=jXUOUVvqWd1yLxfnmXxvB-DiJ44_gfimwQ3bCOKTUpg&e=>
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_DefectDojo_django-2DDefectDojo_discussions_5195-23discussioncomment-2D12697238&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=aNxHlfDsWx5OdvjtsyfULQ1oCWePGPJ4WOrfMYHzu2-UN2lb791PQcnDh5QeqPTr&s=OHe4t_7SKCil2_zOZafYQd5-6T4xfRFYrBS9JFmwajA&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_BCGFMJCM7EUV4V5JMDVHZXL2XOH7FAVCNFSM6AAAAABKQGMJP6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENRZG4ZDGOA&d=DwMFaQ&c=udBTRvFvXC5Dhqg7UHpJlPps3mZ3LRxpb6__0PomBTQ&r=M4cQrVBS_6Q0Bvpv1_nFVRmmM2LHVDx5h50kPTyCuKg&m=aNxHlfDsWx5OdvjtsyfULQ1oCWePGPJ4WOrfMYHzu2-UN2lb791PQcnDh5QeqPTr&s=-OgXgaCWxR5zOQbjk1WZScmrncl29fd2RlLDVDbhJnI&e=>
> .
> You are receiving this because you were mentioned.Message ID:
> <DefectDojo/django-DefectDojo/repo-discussions/5195/comments/12697238@
> github.com>
>


-- 

Click here for "Request Forms"
<https://sites.google.com/verizon.com/publicsectorcso-scan/request-forms>


*Josh Noonan – CISSP, CEH, ECSA*

Public Sector CyberSecurity Organization

Verizon

703-728-2733

**Reply by Noonster77** [reply]
Let me know if its working!

**Reply by stanleylin75** [reply]
thanks, it works. 

but the findings are import to the same product 

![image](https://github.com/user-attachments/assets/35078154-5328-4f72-b8b9-00b8cf2d78c7)
![image](https://github.com/user-attachments/assets/e24195ce-55cf-4071-9626-1039a4d14d48)
![image](https://github.com/user-attachments/assets/24b6bae0-b67a-472d-b7d1-b1a33b1cf9c2)

**Comment by Noonster77** [comment]
I am still seeing it exporting the tenable.sc query with no tags. Make sure the query has tags from tenable.sc and they match the product from defect dojo.
![image](https://github.com/user-attachments/assets/17d31151-c031-4257-abe0-520e7fc24439)
![image](https://github.com/user-attachments/assets/8b0abeed-f21d-4c91-9d14-1d602e578186)
There should be one query per product.

**Reply by stanleylin75** [reply]
it tags in tenable.sc, no idea why it still query without tags.
the product name is also same as the tags.

![image](https://github.com/user-attachments/assets/e576e539-9d7c-4113-a666-11088f9244e2)

![image](https://github.com/user-attachments/assets/118dbb7f-84e3-4821-802b-4e403d1aaff2)

**Reply by Noonster77** [reply]
Its working fine in multiple different environments in my company. The only thing I could think of is the user you are using to export from tenable.sc via the API key may not have permissions to work with all objects. Make sure its a higher access account. We can always jump on a call if needed.
